File: Queue/C/DQueue.c
#include<stdio.h>
#include<stdlib.h>

struct node 
{
    int data;
    struct node * next;
    struct node * prev;
};

typedef struct node NODE;
typedef struct node * PNODE;
typedef struct node ** PPNODE;

void EnqueueFirst(PPNODE first, PPNODE last, int no)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));
    newn->data = no;
    newn->next = NULL;
    newn->prev = NULL;

    if((*first == NULL) && (*last == NULL))
    {
        *first = newn;
        *last = newn;
    }
    else
    {
        newn->next = *first;
        (*first)->prev = newn;

        *first = newn;
    }
}

void EnqueueLast(PPNODE first, PPNODE last, int no)
{
    PNODE newn = NULL;

    newn = (PNODE)malloc(sizeof(NODE));
    newn->data = no;
    newn->next = NULL;
    newn->prev = NULL;

    if((*first == NULL) && (*last == NULL))
    {
        *first = newn;
        *last = newn;
    }
    else
    {
        (*last)->next = newn;
        newn->prev = *last;

        *last = newn;
    }
}

int DequeueFirst(PPNODE first, PPNODE last)
{
    int value = 0;
    PNODE temp = NULL;

    if((*first == NULL) && (*last == NULL))
    {
        printf("Unable to delete Queue is empty\n");
        return -1;
    }
    else if(*first == *last)
    {
        value = (*first)->data;
        
        free(*first);

        *first = NULL;
        *last = NULL;
    }
    else
    {
        temp = *first;

        value = temp->data;

        *first = (*first)->next;
        (*first)->prev = NULL;
        
        free(temp);
    }

    return value;
}

int DequeueLast(PPNODE first, PPNODE last)
{
    int value = 0;
    PNODE temp = NULL;

    if((*first == NULL) && (*last == NULL))
    {
        printf("Unable to delete Queue is empty\n");
        return -1;
    }
    else if(*first == *last)
    {
        value = (*first)->data;
        
        free(*first);

        *first = NULL;
        *last = NULL;
    }
    else
    {
        temp = *last;

        value = temp->data;

        *last = (*last)->prev;
        (*last)->next = NULL;
        
        free(temp);
    }

    return value;
}

void DisplayFirst(PNODE first, PNODE last)
{
    if((first == NULL) && (last == NULL))
    {
        printf("Queue is empty\n");
        return;
    }

    printf("NULL =");
    while (first != NULL)
    {
        printf(" | %d | = ",first->data);
        first = first->next;
    }
    
    printf("NULL\n");
}

void DisplayLast(PNODE first, PNODE last)
{
    if((first == NULL) && (last == NULL))
    {
        printf("Queue is empty\n");
        return;
    }

    printf("NULL =");
    while (last != NULL)
    {
        printf(" | %d | = ",last->data);
        last = last->prev;
    }
    
    printf("NULL\n");
}

int Count(PNODE first, PNODE last)
{
    int iCount = 0;

    if((first == NULL) && (last == NULL))
    {
        printf("Queue is empty\n");
        return 0;
    }

    while (first != NULL)
    {
        iCount++;
        first = first->next;
    }

    return iCount;
}

int main()
{
    PNODE head = NULL;
    PNODE tail = NULL;
    int choice = 0;
    int value = 0;
    int iRet = 0;

    printf("\n=============================================================\n");
    printf("                Doubly Ended Queue (Deque) Manager                    \n");
    printf("=============================================================\n");

    while(1)
    {
        printf("\n-------------------------------------------------------------\n");
        printf("Please select an operation:\n");
        printf("-------------------------------------------------------------\n");
        printf(" 1 : Enqueue at front\n");
        printf(" 2 : Enqueue at rear\n");
        printf(" 3 : Dequeue from front\n");
        printf(" 4 : Dequeue from rear\n");
        printf(" 5 : Display from front to rear\n");
        printf(" 6 : Display from rear to front\n");
        printf(" 7 : Count total elements\n");
        printf(" 0 : Exit\n");
        printf("-------------------------------------------------------------\n");

        printf("Enter your choice: ");
        scanf("%d",&choice);

        if(choice == 1)
        {
            printf("Enter element to enqueue at front: ");
            scanf("%d",&value);
            EnqueueFirst(&head, &tail, value);
        }
        else if(choice == 2)
        {
            printf("Enter element to enqueue at rear: ");
            scanf("%d",&value);
            EnqueueLast(&head, &tail, value);
        }
        else if(choice == 3)
        {
            value = DequeueFirst(&head, &tail);
            printf("Dequeued from front: %d\n",value);
        }
        else if(choice == 4)
        {
            value = DequeueLast(&head, &tail);
            printf("Dequeued from rear: %d\n",value);
        }
        else if(choice == 5)
        {
            DisplayFirst(head, tail);
        }
        else if(choice == 6)
        {
            DisplayLast(head, tail);
        }
        else if(choice == 7)
        {
            iRet = Count(head, tail);
            printf("Total elements in deque: %d\n",iRet);
        }
        else if(choice == 0)
        {
            printf("Thank you for using our Application!\n");
            break;
        }
        else
        {
            printf("Invalid choice. Please try again.\n");
        }
    }
    
    return 0;
}

File: Queue/Python/DQueue.py

class node:

    def __init__(self, no):
        self.data = no
        self.next = None
        self.prev = None

class DQueue:

    def __init__(self):
        self.first = None
        self.last = None
        self.iCount = 0

    def enqueuefirst(self, no):

        newn = node(no)

        if(self.first == None and self.last == None):
            self.first = newn
            self.last = newn

        else:
            newn.next = self.first
            self.first.prev = newn
            self.first = newn

        self.iCount += 1


    def enqueuelast(self, no):

        newn = node(no)

        if(self.first == None and self.last == None):
            self.first = newn
            self.last = newn

        else:
            self.last.next = newn
            newn.prev = self.last

            self.last = newn

        self.iCount += 1


    def dequeuefirst(self):

        if(self.first == None and self.last == None):
            print("Unable to delete Queue is empty")
            return None

        elif(self.first == self.last):
            iValue = self.first.data
            
            del self.first

            self.first = None
            self.last = None
    
        else:
            iValue = self.first.data

            self.first = self.first.next
            del self.first.prev
            self.first.prev = None

        self.iCount -= 1
        return iValue

    def dequeuelast(self):

        if(self.first == None and self.last == None):
            print("Unable to delete Queue is empty")
            return None

        elif(self.first == self.last):
            iValue = self.last.data
            
            del self.last

            self.first = None
            self.last = None
    
        else:
            iValue = self.last.data

            self.last = self.last.prev
            del self.last.next
            self.last.next = None

        self.iCount -= 1
        return iValue

    def Displayfirst(self):

        if(self.first == None and self.last == None):
            print("Queue is empty")
            return

        temp = self.first

        print("None <=> ",end="")

        while(temp != None):
            print(f"| {temp.data} | <=> ",end="");
            temp = temp.next

        print("None")


    def Displaylast(self):

        if(self.first == None and self.last == None):
            print("Queue is empty")
            return

        temp = self.last

        print("None <=> ",end="")

        while(temp != None):
            print(f"| {temp.data} | <=> ",end="");
            temp = temp.prev

        print("None")

    
    def Count(self):
        return self.iCount


def main():

    obj = DQueue()

    print("\n=============================================================");
    print("              Doubly Ended Queue (Deque) Manager                    ");
    print("=============================================================");

    while True:
        print("\n-------------------------------------------------------------")
        print("Please select an operation:")
        print("-------------------------------------------------------------")
        print(" 1 : Enqueue at front");
        print(" 2 : Enqueue at rear");
        print(" 3 : Dequeue from front");
        print(" 4 : Dequeue from rear");
        print(" 5 : Display from front to rear");
        print(" 6 : Display from rear to front");
        print(" 7 : Count total elements");
        print(" 0 : Exit");
        print("-------------------------------------------------------------")

        choice = int(input("Enter your choice : "))

        if choice == 1:
            value = int(input("Enter element to enqueue at front: "))
            obj.enqueuefirst(value)

        elif choice == 2:
            value = int(input("Enter element to enqueue at rear: "))
            obj.enqueuelast(value)

        elif choice == 3:
            iRet = obj.dequeuefirst()
            if iRet is not None:
                print(f"Dequeued from front: {iRet}")

        elif choice == 4:
            iRet = obj.dequeuelast()
            if iRet is not None:
                print(f"Dequeued from rear: {iRet}")

        elif choice == 5:
            obj.Displayfirst()

        elif choice == 6:
            obj.Displaylast()
        
        elif choice == 7:
            print(f"Total elements in Dqueue: {obj.Count()}")

        elif choice == 0:
            print("Thank you for using our Application!")
            break

        else:
            print("Invalid choice, please try again")

if __name__ == "__main__":
    main()